python/ray/train/_internal/storage.py

```diff
@@ -0,0 +1,154 @@
+import dataclasses
+import os
+from typing import Optional
+
+import pyarrow.fs
+
+from ray.air._internal.uri_utils import is_uri
+from ray.air._internal.remote_storage import get_fs_and_path
+from ray.tune.syncer import Syncer, SyncConfig, _DefaultSyncer
+from ray.tune.result import _get_defaults_results_dir
+
+
+def _use_storage_context() -> bool:
+    # Whether to enable the new simple persistence mode.
+    return bool(int(os.environ.get("RAY_AIR_NEW_PERSISTENCE_MODE", "0")))
+
+
+class StorageContext:
+    """Shared context that holds all paths and storage utilities, passed along from
+    the driver to workers.
+
+    Example:
+        storage_path = "s3://bucket/results"
+        storage_filesystem = S3FileSystem (auto-resolved)
+        storage_cache_path = "~/ray_results"
+        storage_fs_path = "bucket/results"
+
+        experiment_dir_name = "exp_name"
+        # experiment_path = "s3://bucket/results/exp_name"
+        experiment_fs_path = "bucket/results/exp_name"  # Use this path with pyarrow
+        experiment_cache_path = "~/ray_results/exp_name"
+
+        # Only set on workers.
+        trial_dir_name = "trial_dir"
+        # trial_path = "s3://bucket/results/exp_name/trial_dir"
+        trial_fs_path = "bucket/results/exp_name/trial_dir"
+        # trial_cache_path = "~/ray_results/exp_name/trial_dir"
+
+    Internal Usage Examples:
+        construct_checkpoint_fs_path("checkpoint_000001")
+        -> "bucket/results/exp_name/trial_dir/checkpoint_000001"
+
+        pyarrow.fs.copy_files(
+            local_dir,
+            os.path.join(storage.trial_fs_path, "subdir"),
+            destination_filesystem=storage.filesystem
+        )
+    """
+
+    def __init__(
+        self,
+        storage_path: str,
+        sync_config: SyncConfig,
+        experiment_dir_name: str,
+        storage_filesystem: Optional[pyarrow.fs.FileSystem] = None,
+        trial_dir_name: Optional[str] = None,
+    ):
+        self.storage_path: str = storage_path
+        self.storage_cache_path: str = _get_defaults_results_dir()
+        self.experiment_dir_name: str = experiment_dir_name
+        self.trial_dir_name: Optional[str] = trial_dir_name
+        self.sync_config: SyncConfig = dataclasses.replace(sync_config)
+
+        if storage_filesystem:
+            # Custom pyarrow filesystem
+            self.storage_filesystem = storage_filesystem
+            if is_uri(self.storage_path):
+                raise ValueError("TODO")
+            self.storage_fs_path = self.storage_path
+        else:
+            (
+                self.storage_filesystem,
+                self.storage_fs_path,
+            ) = get_fs_and_path(self.storage_path)
```

Comment:
I stick with `get_fs_and_path` and the utils in `ray.air._internal.remote_storage`, since there are many fixes in there (ex: hdfs uri parsing, pyarrow bug causing uploads of many files to hang forever, as well as bad performance from pyarrow defaults).

There may be a bug in windows path parsing that needs to be fixed.

Reply:
Let's discuss. I think we need to take a strong stand here, and simplify how this works.

At a high level, Ray Data is working fine with pyarrow filesystem URI resolution, and I don't see why Train needs to diverge here. I doubt the assumptions made before are necessarily valid now.

Reply:
Oh, now that I'm looking at this HDFS URI parsing again, the only reason this is needed is because we cache the pyarrow filesystem and then try to perform the prefix stripping ourselves on the next filesystem operation (using the cached one). We can simplify / remove this.

The one thing that I think we should keep from `get_fs_and_path` is replacing the default s3/gs syncers with the corresponding **fsspec implementations.** See https://github.com/ray-project/ray/pull/34663

In particular, the default pyarrow S3FileSystem has many issues:
1. It has a bug that causes uploads of many files to hang forever when using multiple threads to upload. So, we need to manually limit this to use a single thread.
2. It's slow due to 1 (see the linked PR for performance comparisons).

So, my new imagined `get_fs_and_path`:

```python
if fsspec installed:
    return fsspec s3fs/gs filesystem from URI

return pyarrow.fs.from_uri(uri)
```

Reply:
Yes, I'm aware of the benchmarks. We're talking about what looks like ~40% performance improvement, however, for an operation that is typically infrequent (<1% of training overhead, given reasonably configured checkpoint intervals).

I'd like us to drop the fsspec dependency altogether here and standardize, given that this is also likely to be fixed in the future.